### PR TITLE
adding weights for C++ (#267)

### PR DESCRIPTION
**Pull request by joehellmersNOAA**
## What type of PR is this? (check all applicable)

- [X] Refactor
- [ ] Feature
- [X] Bug Fix
- [ ] Optimization
- [ ] Example
- [ ] Documentation

## Description
This pull request is needed to fix the C++ weights, and does so by separating them into two new classes that are not part of the associated operators.

## Related Issues & Documents

- Closes #
#224 #216 

## QA Instructions, Screenshots, Recordings

_Please replace this line with instructions on how to test your changes, a note
on the devices and browsers this has been tested on, as well as any relevant
images for UI changes._

## Added/updated tests?
_We encourage you to test all code included with MOLE, including examples.

- [X] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## Read Contributing Guide and Code of Conduct

- [X] I have read and followed the [Contributing Guide](https://github.com/csrc-sdsu/mole/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/csrc-sdsu/mole/blob/main/CODE_OF_CONDUCT.md)

## [optional] Are there any post deployment tasks we need to perform?
No

### CODE REVIEW

**Review by jbrzensk — COMMENTED** [review]
Yeah @joehellmersNOAA !!
I put some comments, mostly have to do with spacing, using the arma:: before Armadillo things, and  **const** before the constant declaration in the test.

Good for 1D, I think that 2D and 3D are not as trivial as we think, hence the 2D in MATLAB, and no 3D implementations! I would leave the 2D and 3D until there is a better understanding, and a need, for them.

File: src/cpp/utils.cpp [hunk 1]
```diff
@@ -47,7 +48,34 @@ vec Utils::spsolve_eigen(const sp_mat &A, const vec &b) {
 
   return vec(eigen_x.data(), eigen_x.size());
 }
-#endif
+
+vec Utils::spsolve_eigenQR(const sp_mat &A, const vec &b) {
```
Comment (aboada) [inline src/cpp/utils.cpp]: Could you elaborate on why QR factorization is needed here as opposed to LU? I’m assuming this is for computing P and Q. Are we expecting any ill-conditioned matrices?

**Review by aboada — COMMENTED** [review]
Thanks, still working on the review. I recommend applying clang-format (see: [ClangFormat Documentation](https://clang.llvm.org/docs/ClangFormat.html)) to fix any formatting issues.

File: tests/cpp/test6.cpp
Comment (aboada) [inline tests/cpp/test6.cpp]: Could we give this test a more descriptive name?  Since we may add more functionality in the future, how will we know that test6 is associated with `P` and `Q`?
Changing the test name may incur additional updates to the instructions in the `CMakeList`. If that's too much, we can do it in a separate PR.

File: tests/cpp/test6.cpp [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    for (int i = 0; i < m; ++i) {
+        total_error += abs(P[i] - P_baseline[j][i]);
+    }
+    ASSERT_LE(total_error, tolerance);
+
+}
+
+TEST(WeightTests, Accuracy) {
+
+        // Baseline weights - Generated from MatLab/Octave
```
Comment (aboada) [inline tests/cpp/test6.cpp]: Instead of hardcoding these values from Matlab/Octave, why don't we check that $G^Tp = b_{m+2}$ (and the same for $Q$), as shown in [Corbino and Castillo](https://doi.org/10.1016/j.cam.2019.06.042)?

We can also have a separate test analogous to [examples/matlab_octave/integration1D.m](https://github.com/csrc-sdsu/mole/blob/main/examples/matlab_octave/integration1D.m).

File: tests/cpp/test6.cpp [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        P_baseline[1] = {
+            0.354134518,1.228459404,0.898117099,1.018547095,1.000741884,
+            1.000741884,1.018547095,0.898117099,1.228459404,0.354134518
+        };
+        P_baseline[2] = {
+            0.315722926,1.390677390,0.629532984,1.234237262,0.919144328,1.009804636,1.000880474,
+            1.000880474,1.009804636,0.919144328,1.234237262,0.629532984,1.390677390,0.315722926
+        };
+      
+        for (int k : {2,4,6}) {
```
Comment (aboada) [inline tests/cpp/test6.cpp]: To ensure robustness, we should also add tests that check for unexpected, unsupported, corner cases, etc.

For instance, what happens with negative `k` when calling the `WeightsQ` constructor `(k,m,dx)`. What about odd values? Negative `m` or `dx`, etc.

File: src/cpp/utils.cpp (hunk 1 above)
Comment (joehellmersNOAA) [reply src/cpp/utils.cpp]: Since the computation of the weights involves the solution of a linear system were MxN and M <> N, Eigen/SuperLU wants to use QR decomposition.

Regarding the code duplication I see a few options to combine the routines

1. Modify the solver to check for the shape of the matrix.  If square, then use LU, otherwise use QR
2. Use QR all the time, it should work either way.  It probably isn't that much more inefficient than conditionally doing LU, but we could verify that.
3. Add an optional input parameter to let the caller decide to use QR or LU, but force QR if M <> N

BTW This method has issues when order is higher than 6 which can dealt with using LPP (Linear Programming Problem) that is mentioned in [here](https://github.com/csrc-sdsu/mole/issues/224#issuecomment-3661127623) by @mdumett

**Comment by joehellmersNOAA** [comment]
> Thanks, still working on the review. I recommend applying clang-format (see: [ClangFormat Documentation](https://clang.llvm.org/docs/ClangFormat.html)) to fix any formatting issues.

I have no problem passing this C++ code through some sort of linter, but others should weigh in if this is the best one or not.  @jbrzensk @cpaolini
